In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from scipy.stats import bootstrap
import os

In [2]:
# Define o diretório base como a pasta do projeto
BASE_DIR = os.getcwd()

DATA_PATH = os.path.join(BASE_DIR, '..', 'data', 'geo_data_0.csv')

data_0 = pd.read_csv(DATA_PATH )
print(data_0.head(5))

      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647


In [3]:
print(data_0.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [4]:
# Define o diretório base como a pasta do projeto
BASE_DIR = os.getcwd()

DATA_PATH = os.path.join(BASE_DIR, '..', 'data', 'geo_data_1.csv')

data_1 = pd.read_csv(DATA_PATH )

print(data_1.head(5))

      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305


In [5]:
print(data_1.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [6]:
# Define o diretório base como a pasta do projeto
BASE_DIR = os.getcwd()

DATA_PATH = os.path.join(BASE_DIR, '..', 'data', 'geo_data_2.csv')

data_2 = pd.read_csv(DATA_PATH )

print(data_2.head(5))

      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746


In [7]:
print(data_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [8]:
# Função para treinar o modelo e imprimir resultados
def train_and_evaluate(data):
    # Divida os dados em um conjunto de treinamento e um conjunto de validação em uma proporção de 75:25
    features = data.drop(['id', 'product'], axis=1)
    target = data['product']
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=12345
    )
    
    # Treine o modelo e faça previsões para o conjunto de validação
    model = LinearRegression()
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    
    # Imprimir o volume médio previsto de reservas e o RMSE do modelo
    print(f"Volume médio previsto de reservas: {predictions.mean()}")
    rmse = np.sqrt(mean_squared_error(target_valid, predictions))
    print(f"RMSE do modelo: {rmse}")
    
    # Retornar as previsões e as respostas corretas
    return predictions, target_valid


In [9]:
# Dados do geo_data_0.csv
print('Resultados para o arquivo geo_data_0.csv:')
predictions_0, target_valid_0 = train_and_evaluate(data_0)

# Dados do geo_data_1.csv
print('\nResultados para o arquivo geo_data_1.csv:')
predictions_1, target_valid_1 = train_and_evaluate(data_1)

# Dados do geo_data_2.csv
print('\nResultados para o arquivo geo_data_2.csv:')
predictions_2, target_valid_2 = train_and_evaluate(data_2)

# Adicionando o código para criar a 'predictions_list'
predictions_list = [predictions_0, predictions_1, predictions_2]

# Adicionando o código para criar a 'target_list'
target_list = [target_valid_0, target_valid_1, target_valid_2]

Resultados para o arquivo geo_data_0.csv:
Volume médio previsto de reservas: 92.59256778438035
RMSE do modelo: 37.5794217150813

Resultados para o arquivo geo_data_1.csv:
Volume médio previsto de reservas: 68.72854689544602
RMSE do modelo: 0.8930992867756165

Resultados para o arquivo geo_data_2.csv:
Volume médio previsto de reservas: 94.96504596800489
RMSE do modelo: 40.02970873393434


Esses resultados mostram o desempenho dos modelos de regressão linear em prever o volume de reservas de petróleo. A região 1 (geo_data_1.csv) apresentou o menor erro na previsão (RMSE), sugerindo que o modelo foi capaz de prever os volumes de reserva com maior precisão nessa região.

In [10]:
# Definindo as variáveis necessárias para os cálculos
BUDGET_PER_WELL = 100_000_000 / 200  # Orçamento por poço
REVENUE_PER_BARREL = 4500  # Receita por barril em dólares
BREAK_EVEN_VOLUME_PER_WELL = 500_000 / REVENUE_PER_BARREL  # Volume em barris necessários por poço para evitar prejuízos

# Comparando essa quantidade com o volume médio de cada região
regions = ['Região 0', 'Região 1', 'Região 2']
average_volumes = [predictions_0.mean(), predictions_1.mean(), predictions_2.mean()]

# Exibindo as conclusões sobre a etapa de preparação para o cálculo de lucro
for i, avg_volume in enumerate(average_volumes):
    print(f"{regions[i]}: Volume médio previsto de reservas é {avg_volume:.2f} milhares de barris")
    print(f"  - Volume necessário para evitar prejuízos: {BREAK_EVEN_VOLUME_PER_WELL:.2f} milhares de barris")
    
    if avg_volume >= BREAK_EVEN_VOLUME_PER_WELL:
        print("  - Esta região possui um volume médio acima do ponto de equilíbrio e é lucrativa.")
    else:
        print("  - Esta região possui um volume médio abaixo do ponto de equilíbrio e pode não ser lucrativa.")
    print()

Região 0: Volume médio previsto de reservas é 92.59 milhares de barris
  - Volume necessário para evitar prejuízos: 111.11 milhares de barris
  - Esta região possui um volume médio abaixo do ponto de equilíbrio e pode não ser lucrativa.

Região 1: Volume médio previsto de reservas é 68.73 milhares de barris
  - Volume necessário para evitar prejuízos: 111.11 milhares de barris
  - Esta região possui um volume médio abaixo do ponto de equilíbrio e pode não ser lucrativa.

Região 2: Volume médio previsto de reservas é 94.97 milhares de barris
  - Volume necessário para evitar prejuízos: 111.11 milhares de barris
  - Esta região possui um volume médio abaixo do ponto de equilíbrio e pode não ser lucrativa.



*Conclusões da etapa de preparação:* Nenhuma das regiões atende ao critério de produzir um volume médio de petróleo que corresponda ao ponto de equilíbrio financeiro, considerando o orçamento e a receita por barril estabelecidos. Isso sugere que, sob as condições atuais, todas as regiões podem representar um risco financeiro e exigirão uma análise de viabilidade mais aprofundada para determinar se o investimento pode ser justificado com base em outros fatores ou estratégias de mitigação de risco.

In [11]:
def calculate_profit_based_on_predictions(predictions, target, budget_per_well, revenue_per_barrel, break_even_volume_per_well):
    # Selecionar os 200 melhores poços
    top_predictions_indices = predictions.argsort()[-200:][::-1]
    top_predictions = predictions[top_predictions_indices]
    top_targets = target.iloc[top_predictions_indices].values
    
    # Sumarizar o volume total das reservas para os 200 melhores poços
    total_volume = top_targets.sum()
    
    # Calcular o lucro total para os 200 melhores poços
    total_revenue = total_volume * revenue_per_barrel
    total_profit = total_revenue - (budget_per_well * 200)
    
    return total_profit, total_volume

# Calcular o lucro potencial para cada região
profits = []
volumes = []

for i, predictions in enumerate(predictions_list):
    profit, volume = calculate_profit_based_on_predictions(predictions, target_list[i], BUDGET_PER_WELL, REVENUE_PER_BARREL, BREAK_EVEN_VOLUME_PER_WELL)
    profits.append(profit)
    volumes.append(volume)
    print(f"{regions[i]}:")
    print(f"  - Volume total de reservas dos 200 melhores poços: {volume:.2f} milhares de barris")
    print(f"  - Lucro potencial: ${profit:,.2f}")

# Sugerir a região para desenvolvimento
best_region_index = np.argmax(profits)
best_region_profit = profits[best_region_index]
best_region_volume = volumes[best_region_index]
print(f"\nA região sugerida para desenvolvimento de poços de petróleo é a {regions[best_region_index]} com um lucro potencial de ${best_region_profit:,.2f} e um volume total de {best_region_volume:.2f} milhares de barris.")

Região 0:
  - Volume total de reservas dos 200 melhores poços: 29601.84 milhares de barris
  - Lucro potencial: $33,208,260.43
Região 1:
  - Volume total de reservas dos 200 melhores poços: 27589.08 milhares de barris
  - Lucro potencial: $24,150,866.97
Região 2:
  - Volume total de reservas dos 200 melhores poços: 28245.22 milhares de barris
  - Lucro potencial: $27,103,499.64

A região sugerida para desenvolvimento de poços de petróleo é a Região 0 com um lucro potencial de $33,208,260.43 e um volume total de 29601.84 milhares de barris.


A análise dos lucros potenciais e volumes totais de reservas dos 200 melhores poços de petróleo em três regiões distintas sugere que a Região 0 seria a escolha mais promissora para o desenvolvimento de poços de petróleo, considerando que esta região possui o maior lucro potencial de $33,208,260.43 e também o maior volume total de reservas dos 200 melhores poços, com 29601.84 milhares de barris.

No entanto, é crucial observar que, apesar da Região 0 apresentar o melhor desempenho entre as três, todas as regiões exibem um lucro potencial que indica uma estratégia de desenvolvimento atualmente não viável financeiramente. Isso é evidenciado pelo fato de que, mesmo para a região mais lucrativa, o lucro potencial está marginalmente acima do orçamento de desenvolvimento d $100 milhões, com um lucro adicional de apenas cerca d $33 milhões. Além disso, os volumes totais de reservas estão consideravelmente acima do volume de equilíbrio necessário para evitar prejuízos, indicando que, apesar de não recomendável, a Região 0 tem a melhor chance de retorno sobre o investimento com base nos critérios estabelecidos.

Esses resultados indicam que uma revisão da estratégia de desenvolvimento atual é necessária. Tal revisão poderia incluir a redução de custos de desenvolvimento, o aumento da receita por barril, ou a busca por métodos que melhorem a precisão das previsões para identificar poços mais produtivos. Adicionalmente, outros fatores não quantificados neste modelo, como riscos geológicos ou operacionais, deveriam ser considerados antes de tomar uma decisão de investimento final. Essa análise enfatiza a importância de uma avaliação de risco abrangente e consideração cuidadosa dos potenciais retornos financeiros antes de prosseguir com o desenvolvimento de recursos naturais em qualquer uma das regiões estudadas.os

In [12]:
# Função para calcular riscos e lucros
def calculate_risks_and_profits(data, revenue_per_barrel, budget_per_well, n_samples=1000):
    # Separar os recursos e alvos
    features = data.drop(['id', 'product'], axis=1)
    target = data['product']
    
    # Dividir os dados
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=12345
    )
    
    # Treinar o modelo
    model = LinearRegression()
    model.fit(features_train, target_train)
    
    # Fazer previsões
    predictions = model.predict(features_valid)
    
    # Selecionar os 200 melhores poços
    top_predictions_indices = predictions.argsort()[-200:]
    
    # Amostragem de bootstrap para calcular a distribuição do lucro
    profits = []
    for _ in range(n_samples):
        sampled_indices = np.random.choice(top_predictions_indices, size=200, replace=True)
        sampled_profits = target_valid.iloc[sampled_indices].sum() * revenue_per_barrel - budget_per_well * 200
        profits.append(sampled_profits)
    
    profits = np.array(profits)
    mean_profit = profits.mean()
    lower = np.percentile(profits, 2.5)
    upper = np.percentile(profits, 97.5)
    risk_of_loss = np.mean(profits < 0) * 100
    
    return mean_profit, lower, upper, risk_of_loss

# Calculando riscos e lucros para cada região
results = {}
for i, data in enumerate([data_0, data_1, data_2]):
    mean_profit, lower, upper, risk_of_loss = calculate_risks_and_profits(
        data, REVENUE_PER_BARREL, BUDGET_PER_WELL
    )
    results[i] = {
        'mean_profit': mean_profit,
        '95%_confidence_interval': (lower, upper),
        'risk_of_loss': risk_of_loss
    }

# Apresentando os resultados e conclusões
for i in range(3):
    print(f"Região {i}:")
    print(f"Lucro médio: ${results[i]['mean_profit']:,.2f}")
    print(f"Intervalo de confiança de 95%: (${results[i]['95%_confidence_interval'][0]:,.2f}, ${results[i]['95%_confidence_interval'][1]:,.2f})")
    print(f"Risco de prejuízo: {results[i]['risk_of_loss']:.2f}%\n")

# Determinar a melhor região para desenvolvimento de poços de petróleo
best_region = min(results, key=lambda x: (results[x]['risk_of_loss'], -results[x]['mean_profit']))
print(f"A região sugerida para o desenvolvimento de poços de petróleo é a Região {best_region},")
print(f"com um lucro médio de ${results[best_region]['mean_profit']:,.2f},")
print(f"um intervalo de confiança de 95% de (${results[best_region]['95%_confidence_interval'][0]:,.2f}, ${results[best_region]['95%_confidence_interval'][1]:,.2f}),")
print(f"e um risco de prejuízo de {results[best_region]['risk_of_loss']:.2f}%.")
print(f"Essa escolha leva em conta tanto o potencial de lucro quanto o risco associado.")

Região 0:
Lucro médio: $33,256,343.55
Intervalo de confiança de 95%: ($29,923,454.07, $36,183,226.04)
Risco de prejuízo: 0.00%

Região 1:
Lucro médio: $24,150,866.97
Intervalo de confiança de 95%: ($24,150,866.97, $24,150,866.97)
Risco de prejuízo: 0.00%

Região 2:
Lucro médio: $27,019,016.10
Intervalo de confiança de 95%: ($23,209,269.25, $30,623,363.76)
Risco de prejuízo: 0.00%

A região sugerida para o desenvolvimento de poços de petróleo é a Região 0,
com um lucro médio de $33,256,343.55,
um intervalo de confiança de 95% de ($29,923,454.07, $36,183,226.04),
e um risco de prejuízo de 0.00%.
Essa escolha leva em conta tanto o potencial de lucro quanto o risco associado.


Análise de Lucro e Risco por Regiãordas.

Região 0:

Com um lucro médio de 33,201,960.97 e um intervalo de confiança de 95% que varia de 30,183,737.90 a 36,253,312.84, a Região 0 apresenta-se como a mais promissora para investimento em desenvolvimento de poços de petróleo.
O risco de prejuízo para esta região é de 0.00%, indicando uma alta confiança nas previsões de que o investimento resultará em lucro.
A Região 0 é, portanto, a escolha preferencial para o desenvolvimento devido ao seu maior potencial de retorno financeiro e à ausência de risco de perdas.

Região 1:

A Região 1 mostra um lucro médio d $24,150,866.97. Intrigantemente, o intervalo de confiança de 95% não varia, indicando uma previsão de lucro extremamente consistente e sem variabilidade.
Assim como a Região 0, a Região 1 tem um risco de prejuízo de 0.00%, o que sugere uma certeza de lucro com base nas previsões do modelo.
Apesar do risco de prejuízo ser inexistente, o lucro médio mais baixo comparado à Região 0 significa que o retorno sobre o investimento é menor, tornando-a uma opção secundária.

Região 2:

A Região 2 tem um lucro médio estimado e $27,055,445.91, com um intervalo de confiança de 95% entr $23,282,460.25  $30,726,180.61, o que indica uma maior incerteza nos resultados previstos em comparação com as outras regiões.
O risco de prejuízo é novamente reportado como 0.00%, o que é consistente com as outras regiões, sugerindo que as previsões do modelo são otimistas quanto à rentabilidade do investimento.
A maior variabilidade no lucro potencial, no entanto, coloca a Região 2 como uma opção menos atrativa do que a Regiã 0.

*Conclusão geral:*
As análises das três regiões indicam que todas elas têm um potencial lucrativo com um risco de prejuízo nulo, de acordo com o modelo de previsão utilizado. A Região 0, com o maior lucro médio e um intervalo de confiança que sugere um sólido retorno sobre o investimento, é a escolha mais atraente para o desenvolvimento de poços de petróleo.

É importante notar que um risco de prejuízo reportado como 0.00% em todas as regiões é extraordinário e levanta questões sobre a possibilidade de sobreajuste (overfitting) do modelo ou outras formas de viés nas previsões. Na prática, um investimento sem risco de prejuízo é altamente improvável, especialmente em indústrias com variáveis tão imprevisíveis como a exploração de petróleo. Portanto, é crucial realizar uma análise de risco mais aprofundada para validar essas previsões antes de se comprometer com decisões de investimento significativas.